src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -648,7 +648,13 @@ private BoundCollectionExpression ConvertCollectionExpression(
                 Debug.Assert((collectionCreation is BoundNewT && !isExpanded && constructor is null) ||
                              (collectionCreation is BoundObjectCreationExpression creation && creation.Expanded == isExpanded && creation.Constructor == constructor));
 
-                var collectionInitializerAddMethodBinder = this.WithAdditionalFlags(BinderFlags.CollectionInitializerAddMethod);
+                if (!elements.IsDefaultOrEmpty && HasCollectionInitializerTypeInProgress(syntax, targetType))
+                {
+                    diagnostics.Add(ErrorCode.ERR_CollectionInitializerInfiniteChainOfAddCalls, syntax, targetType);
```

Comment:
> Also, we should confirm that there is no way to get into a cycle when constructors and `Add` methods are involved at the same time, otherwise we might be missing a guard.

I didn't think of a way that would be possible without hitting the existing guards. I can add some more tests though.

> It feels like we should unify and combine both checks, i.e. do them in the same place and use similar approaches for both.

The constructor check is separate and happens before binding the conversion - I considered implementing the Add check at the same place but that would require duplicating the complicated Add binding logic. The other way around (moving the existing constructor check) could be probably done - but it seems orthogonal to this bug fix, perhaps we can just file an issue. Also I think we would need another intermediate binder so it wouldn't share code with the Add check, i.e., maybe there are no advantages.

Reply:
> perhaps we can just file an issue

I think this should  be sufficient for now.

Reply:
https://github.com/dotnet/roslyn/issues/75024